autoload/slime.vim

```diff
@@ -74,7 +74,38 @@ function! s:ZellijSend(config, text)
   if a:config["relative_pane"] != "current"
     call system("zellij " . target_session . " action move-focus " . shellescape(a:config["relative_pane"]))
   end
-  call system("zellij " . target_session . " action write-chars " . shellescape(a:text))
+  if exists("b:slime_bracketed_paste")
+    let bracketed_paste = b:slime_bracketed_paste
+  elseif exists("g:slime_bracketed_paste")
+    let bracketed_paste = g:slime_bracketed_paste
+  else
+    let bracketed_paste = 0
+  endif
+
+  let [text_to_paste, has_crlf] = [a:text, 0]
+  if bracketed_paste
+    if a:text[-2:] == "\r\n"
+      let [text_to_paste, has_crlf] = [a:text[:-3], 1]
+    elseif a:text[-1:] == "\r" || a:text[-1:] == "\n"
+      let [text_to_paste, has_crlf] = [a:text[:-2], 1]
+    endif
+  endif
+
+  if bracketed_paste
+    " zellij doesn't support bracketed paste so wrap in escape codes
+    let ec_start = [27, 91, 50, 48, 48, 126] " ESC [ 2 0 0 ~
+    let ec_end = [27, 91, 50, 48, 49, 126] " ESC [ 2 0 1 ~
+    if has_crlf
+      " trailing newline
+      let ec_end += [10]
+    end
+    " convert to list of ascii codes and stack the escape codes
+    let text_to_paste = ec_start + map(split(text_to_paste, '\zs'), 'char2nr(v:val)') + ec_end
```

Comment:
if the goal is to generate a string with special characters, it might be simpler to use `\x##` inside strings, and prepend/append those strings

some example for elsewhere in the code:
- [scala ftplugin](https://github.com/jpalardy/vim-slime/blob/main/ftplugin/scala/slime.vim#L7)
- [coffescript ftplugin](https://github.com/jpalardy/vim-slime/blob/main/ftplugin/coffee/slime.vim#L4)

Reply:
you're right the escape sequences can go straight to strings - something like this?
```
 if bracketed_paste
    " convert to list of ascii codes and add the escape codes
    let text_to_paste = join(map(split(text_to_paste, '\zs'), 'char2nr(v:val)'))
    let ec_start = "27 91 50 48 48 126 " 
    if has_crlf
      " trailing newline
      let ec_end = " 27 91 50 48 49 126 10" 
    else
      let ec_end = " 27 91 50 48 49 126"  
    endif
    call system("zellij " . target_session . " action write " . ec_start . text_to_paste . ec_end)
  else
    call system("zellij " . target_session . " action write-chars " . shellescape(text_to_paste))
  endif
```
(it seems that the zellij write action only accepts a sequence of base-10 values)

Reply:
I meant to go one step further and skip the join-map-split

something like:

```vim
let ec_start = "\x27[200~"
let ec_end = "\x27[201~"
if has_crlf
  let ec_end = ec_end . "\n"
endif
```

Reply:
the main issue is that the only way it appears to work correctly is by using the zellij action `write` rather than `write-chars` (write sends an array of bytes, whereas write-chars sends a string of characters).

so this works
```
zellij action write 27 91 50 48 48 126 104 101 108 108 111 10 104 101 108 108 111 10 27 91 50 48 49 126 10
```
but this doesn't
```
zellij action write-chars  "\x27[200~hello\nhello\x27[201~\n"
```
With your code we'd still have to include the join-map-split step to get the sequence of digits needed for the call to write. Maybe there's a way to do this using write-chars but I haven't been able to figure it out and I suspect it may be impossible because of the way zellij implements the two commands

Reply:
damn 🤔 

as [it stands now](https://github.com/jpalardy/vim-slime/blob/main/autoload/slime.vim#L69-L81), it seems like the chars will be properly escaped for the shell.

and `write-chars` expects … a unicode string? and there's no loopholes for escape sequences?

meanwhile `write` probably hits some command-line argument limit relatively quickly 😬 

if there's no workaround, I think it's better to sandwich the code with the bracketedpaste sequences (with `if` no less), and keep the original `write-chars` for the real text?

Reply:
right - just updated. I thought I'd tried that before (write then write-chars then write again) and it didn't work but seems ok now so at least it gets rid of the join-map-split and the command line limit.